ITE/4/README.adoc

```diff
@@ -0,0 +1,462 @@
+    },
+    "products": {
+        "github+https://github.com/in-toto/in-toto/commit/f1c5d201887e226cadac5792a203ac3eae347add/checks?check_suite_id=<ID>": <HASH>
+    },
+    "byproducts": {
+        "stderr": "",
+        "stdout": "",
+        "return-value": ""
+    },
+    "environment": {
+        "variables": "",
+        "filesystem": "",
+        "workdir": ""
+    }
+}
+```
+
+[[reasoning]]
+== Reasoning
+
+This ITE proposes a change in the URI scheme of artifacts in in-toto metadata.
+The proposed change closely matches the URI structure laid out in RFC 3986. The
+URI structure proposed in that document is widely accepted and is versatile
+enough to allow for a wide variety of references. It is also easy to implement
+and extend support for due to the presence of a large number of standard
+libraries.
+
+[[backwards-compatibility]]
+== Backwards Compatibility
+
+If in-toto metadata is generated using an implementation of in-toto conforming
+to this ITE, verification using a non-conforming implementation can fail. It is
+possible for the verification workflow to progress if the inspections don't
+use generic URIs.
+
+However, a conforming implementation should be capable of verifying in-toto
+metadata generated using a non-conforming implementation, as an ITE-4 conforming
+system MUST also conform to the actual in-toto specification.
+
+It's also possible two conforming systems may be unable to verify the other's
+in-toto metadata as they're unaware of how to resolve certain URI tokens used by
+the other. This is again because of the possibility of inspections containing
+URIs that the other system is unable to resolve and calculate cryptographic
+information for.
+
+[[security]]
+== Security
+
+It's important to note that file paths are a form of URIs too, and can be
+represented with the `scheme` `file://`. We have also discovered that `MATCH`
+and other artifact rules use only the URI strings themselves to find the
+appropriate entries in the link metadata. The changes proposed in this ITE,
+therefore, have no effect on the security guarantees made by the in-toto
+specification.
+
+Implementing systems that conform to this ITE must, however, be careful with how
+they compute cryptographic hashes of abstract entities. The implementors must
+understand the abstract entities they are hashing so as to be able to do so
+consistently.
+
+[[infrastructure-requirements]]
+== Infrastructure Requirements
+
+This ITE proposes no infrastructure changes.
+
+[[testing]]
+== Testing
+
+In an in-toto system conforming to this ITE, it is important to test:
+
+- that all the artifact rules behave as described in the specification
+- that cryptographic hashes of data in abstract entities change with changes in
+in the data - in effect, this would test how the data is transformed into a form
+that can be hashed
+
+[[prototype-implementation]]
+== Prototype Implementation
+
+This ITE currently proposes no prototypes.
```

Comment:
I think that this is needed for sure.  I'd like to see how this works in practice and have several implementations independently done of the same URI type to see how similar the hashes, etc. are.

Reply:
I think there will be experimental implementations soon because of the ongoing integration with SPDX. Santiago and I will sketch stuff out and I'll update this section with details.

Reply:
I especially want to see a working, stable implementation for GitHub that has been tested for at least a few months. Right now, the ideas sound good, but I'd like to see some empirical validation.